webdriver api code (Fixes#7)

lmalakhova opened the pull request:
Added the following:
-Create simple test using WebDriver API.
-Move driver initialization / cleanup staff outside test, using listeners for that.

lmalakhova committed:
Changed the following:
 - Create simple test using WebDriver API.
 - Move driver initialization / cleanup staff outside tests using listeners.

sskorol commented on src/test/java/com/waverley/InitializationTest.java:
```diff
@@ -0,0 +1,31 @@
+public class InitializationTest extends BaseTest {
+    @BeforeMethod
+    public void checkTestStatus(Method method) {
+        int openTestquantity = 0;
```
check name conventions for your local variables

sskorol commented on src/test/java/com/waverley/listeners/TestListener.java:
```diff
@@ -0,0 +1,55 @@
+ */
+@SuppressWarnings("all")
+public class TestListener implements ITestListener {
+    public static WebDriver driver;
```
try to make driver thread local

sskorol commented on src/test/java/com/waverley/testcases/GoogleTests.java:
```diff
@@ -0,0 +1,22 @@
+    @Test
+    public void shouldSearchForKeyword() throws InterruptedException {
+        driver.get("https://google.com/ncr");
+        final WebElement element = driver.findElement(By.cssSelector("#lst-ib"));
```
use PageObject and refactor corresponding test

lmalakhova committed:
Changed the following:
-Move common configuration into properties.
-Use Owner library for properties handling.
-Move common PageObject actions into BasePage.
-Create generic waiting method within BasePage.
-Increase readability by ExpectedConditions encapsulation within enum.
-Add missing enum values to cover more complicated waiting scenarios.
-Create generic PageFactory for new PageObjects initialization.

sskorol commented on src/main/java/com/waverley/BaseConfig.java:
```diff
@@ -0,0 +1,18 @@
+    long waitTimeout();
+
+    @Key("url")
+    @DefaultValue("http:localhost")
```
protocol should be followed by `://`

sskorol commented on src/main/java/com/waverley/WaitCondition.java:
```diff
@@ -0,0 +1,30 @@
+public enum WaitCondition {
+    visible(ExpectedConditions::visibilityOfElementLocated),
+    present(ExpectedConditions::presenceOfElementLocated),
+    //enabled(ExpectedConditions::elementToBeClickable),
```
commented code should be removed... or either fix an issue with ambiguous types

sskorol commented on src/main/java/com/waverley/WaitCondition.java:
```diff
@@ -0,0 +1,30 @@
+    allvisible(ExpectedConditions::visibilityOfAllElementsLocatedBy),
+    textpresent(ExpectedConditions::textToBePresentInElementLocated),
+    textpresentinvalue(ExpectedConditions::textToBePresentInElementValue);
+    private BiFunction<By, String, ExpectedCondition<?>> type;
```
separate enum keys and instance variable with a new line to increase readability of your code

sskorol commented on src/main/java/com/waverley/WaitCondition.java:
```diff
@@ -0,0 +1,30 @@
+    textpresentinvalue(ExpectedConditions::textToBePresentInElementValue);
+    private BiFunction<By, String, ExpectedCondition<?>> type;
+
+    private WaitCondition(Function<By, ExpectedCondition<?>> type) {
```
missing `final` keyword for both constructors... you may want to run quality plugin to check all issues...

sskorol commented on src/main/java/com/waverley/WaitCondition.java:
```diff
@@ -0,0 +1,30 @@
+    }
+}
+
+
```
seems like a redundant newline

sskorol commented on src/main/java/com/waverley/WebDriverListener.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Class was implemented for testing with Listeners.
+ */
+@SuppressWarnings("all")
```
I don't believe you want to suppress everything

sskorol commented on src/main/java/com/waverley/WebDriverListener.java:
```diff
@@ -0,0 +1,44 @@
+                try {
+                    DRIVER_CONTANER.set(new ChromeDriver());
+                } catch (Exception ex) {
+                    ex.printStackTrace();
```
it's not recommended to print unwrapped stacktrace into console... use logger for that

sskorol commented on src/main/java/com/waverley/WebDriverListener.java:
```diff
@@ -0,0 +1,44 @@
+    @Override
+    public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
+        if (method.isTestMethod()) {
+            ofNullable(getDriver()).ifPresent(d -> d.quit());
```
I believe lambda could be replaced with method reference

sskorol commented on src/main/java/com/waverley/utils/ElementTypeUtils.java:
```diff
@@ -0,0 +1,23 @@
+    public static <T> StreamEx<WebElement> streamof(final T element){
+        return StreamEx.of(listof(element));
+    }
+
```
seems like a redundant new line

sskorol commented on src/main/resources/config.properties:
```diff
@@ -0,0 +1,3 @@
+# WebDriver settings
+explicit.wait.timeout = 5
+url = https://google.com/ncr
```
missing new line

sskorol commented on src/test/java/com/waverley/BaseTest.java:
```diff
@@ -0,0 +1,24 @@
+ * Base class was implemented for Test processing.
+ */
+public abstract class BaseTest {
+    public static final Logger LOG = Logger.getLogger(BaseTest.class.getName());
```
use lombok logger instead of a common one

sskorol commented on src/test/java/com/waverley/testcases/GoogleTests.java:
```diff
@@ -1,22 +1,23 @@
-        element.sendKeys("Testing" + Keys.ENTER);
-        assertThat(driver.findElements(By.cssSelector("#rso h3>a")).size()).isEqualTo(14);
+        open(SearchPage.class)
+                .search("Testing" + Keys.ENTER);
```
move Enter key pressing into Page Object layer

sskorol commented on src/test/java/com/waverley/testcases/GoogleTests.java:
```diff
@@ -1,22 +1,23 @@
+                .search("Testing" + Keys.ENTER);
+
+        assertThat(at(SearchPage.class).getLinksAmount()).isEqualTo(14);
         Thread.sleep(5000);
```
avoid sleeps in a final code version

sskorol reviewed (changes requested):
@lmalakhova check comments please
